### ISSUE TITLE
SDKs should generate files in current working directory by default, not ./dagger

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#6963]
## Problem

When initializing an SDK with `dagger develop` or `dagger init --sdk`, source files are generated in a subdirectory called `./dagger` by default. It seems that most users are either confused by that default, or actively dislike it. I personally always use `--source .`. I would prefer for `.` to be the default, and I think most users will agree.

## Solution

Change the default for `dagger develop --source` and `dagger init --source` to `.` instead of `./dagger`.

## Context

There is actually a reason for the current behavior:

- We encourage users to create a Dagger module at the root of their application repo.
- We also encourage users to (eventually) develop their CI logic as functions in that top-level repo. For this, they use `dagger develop` or `dagger init --sdk`.
- In this situation, it is often preferable to tuck away the generated source code in a subdirectory, instead of polluting the repo root. Thus, the choice of `./dagger`
- However, in all other cases, `.` makes more sense

Nnote: I initially advocated for `./dagger` as the default, so I'll take the L 😅

**Comment by shykes** [comment]
FYI @sipsma @samalba @jpadams

**Comment by nipuna-perera** [comment]
I wish I had more emojis to express how much I like this :)

**Comment by helderco** [comment]
It's actually worse for me because I never init on the current dir, I always do it in a new subdir:
```
❯ dagger init --sdk python --source foo foo
```

Small bikeshedding... due to the above, default should be `""` rather than `"."`, falling back to the value of `PATH` because:
```
❯ dagger init --sdk python --source . foo
Error: [...]: source subdir path ".." escapes context
```

**Comment by helderco** [comment]
Reposting:

> I'd like to raise one aspect that might be important to consider also. When you initialize a dagger module in a project with the same language as the SDK, it'll overwrite some files and potentially break the project in some ways. 
>
> For example in the case of go, if I have a `main.go` and a `go.sum` file and I do `dagger init --sdk go --source .` to create a module for this project, I'll completely overwrite the `go.mod` file and it'll also break my project since it'll create a new `main` function within my main package. Similar thing happens for the other SDK's in some particular cases. 
>
> I guess we need to decide what to do in case this happens.

\- _Originally posted by @marcosnils in DEV-3707, 3 days ago_

**Comment by helderco** [comment]
> IMO, I think we should error out if `--source` points to a non-empty directory. If we change the default to `--source=.` (or `""`), then if the current directory is non-empty, then this would error out as well.

This seems like a good idea at first glance but it breaks a few use cases. The more obvious one is when you `dagger init` (without sources) and `dagger develop --sdk` afterwards.

Additionally, Python and TypeScript allow for configuring the runtime container via `pyproject.toml` and `package.json`. So you can bootstrap your files using your own template (e.g., `rye init`), and `dagger develop --sdk=` should work with those files.

However, in Python for example, if there's a `requirements.lock` file it'll be updated. It's the same that happens with `go.mod`.

**Comment by helderco** [comment]
Maybe a prompt to ask for confirmation? With a `-f` or `-y` flag to force?

**Comment by sagikazarmark** [comment]
I think it would make sense to distinguish between different use cases.

I definitely don't want to see dagger files in the root of my software projects. I often call these directories `ci`, but I started to use the default `dagger` these days.

Another use case I have is my daggerverse repo: I usually create a new directory for a new module and then run `dagger init --source .`.

Maybe it would make sense to have a layout option (Dagger module, project, etc) that sets different defaults. Or even an interactive wizard. It would probably help newcomers decide what to do.

**Comment by marcosnils** [comment]
> (Update: BTW it seems to me that `dagger init` doesn't generate a separate go.mod anymore, but messes with the root go.mod which actively drives me mad)

is that so? :thinking: 

If I do: 
```shell
1|marcos:tmp/test (⎈ |kind-kind)$ go mod init test
go: creating new go.mod: module test
marcos:tmp/test (⎈ |kind-kind)$ dagger i^C
130|marcos:tmp/test (⎈ |kind-kind)$ ls
go.mod
marcos:tmp/test (⎈ |kind-kind)$ ce^C
130|marcos:tmp/test (⎈ |kind-kind)$ dagger init --sdk go
17:10:02 WRN no LICENSE file found; generating one for you, feel free to change or remove license=Apache-2.0
Initialized module test in .
marcos:tmp/test (⎈ |kind-kind)$ tree .
.
├── dagger
│   ├── dagger.gen.go
│   ├── go.mod
│   ├── go.sum
│   ├── internal
│   │   ├── dagger
│   │   │   └── dagger.gen.go
│   │   ├── querybuilder
│   │   │   ├── marshal.go
│   │   │   └── querybuilder.go
│   │   └── telemetry
│   │       ├── attrs.go
│   │       ├── batch_processor.go
│   │       ├── init.go
│   │       ├── processor.go
│   │       ├── proxy.go
│   │       └── span.go
│   └── main.go
├── dagger.json
├── go.mod
└── LICENSE
```

it doesn't seem to messing with my main go.mod file. It creates a new one under the `dagger` folder.

**Comment by helderco** [comment]
I've pinned it down to Go's codegen binary, when you have .go files in the root of a .git repo.

**Comment by miend** [comment]
I don't have a solution suggestion, but throwing my opinion in along with OP -- if I `git init`, it works on the current directory. If I `go mod init`, it works on the current directory. If I `dagger init --sdk=go`, I expect it to work on the current directory. :thinking:

**Comment by shykes** [comment]
> I don't have a solution suggestion, but throwing my opinion in along with OP -- if I `git init`, it works on the current directory, making this directory a git repo. If I `go mod init`, it works on the current directory, turning this directory into a go module. If I `dagger init --sdk=go`, I expect it to work on the current directory, turning this directory into a dagger module. 

To be clear, that is what `dagger init does`. There are two different directories:

1) the module's root directory, which contains `dagger.json`
2) the module's "source directory" which is where the module's SDK reads and writes files.

**Comment by miend** [comment]
> > I don't have a solution suggestion, but throwing my opinion in along with OP -- if I `git init`, it works on the current directory, making this directory a git repo. If I `go mod init`, it works on the current directory, turning this directory into a go module. If I `dagger init --sdk=go`, I expect it to work on the current directory, turning this directory into a dagger module.
> 
> To be clear, that is what `dagger init does`. There are two different directories:
> 
> 1. the module's root directory, which contains `dagger.json`
> 2. the module's "source directory" which is where the module's SDK reads and writes files.

I suppose I could have been clearer -- I expected the source directory to be the current one I'm in. I don't have a hard reason why it should be that way, but that's what I naturally expect when not explicitly specifying a directory name to use.

**Comment by shykes** [comment]
> I suppose I could have been clearer -- I expected the source directory to be the current one I'm in. I don't have a hard reason why it should be that way, but that's what I naturally expect when not explicitly specifying a directory name to use.

Agreed :)

**Comment by sagikazarmark** [comment]
> I suppose I could have been clearer -- I expected the source directory to be the current one I'm in. I don't have a hard reason why it should be that way, but that's what I naturally expect when not explicitly specifying a directory name to use.

I'm starting to believe that the lack of differentiation between a reusable module and a module that implements CI/CD (or other workflows) is the source of a lot of confusion.

Those two are completely different use cases IMO:

For a reusable module, I tend to agree with you. For these cases, git and go are good analogies.


For a non-reusable module, this behavior is not desirable IMO. A better analogy for this scenario would be your CI pipeline files (currently residing in directories like .github and .circle). Would you want those in your project root? Probably not.

**Comment by marcosnils** [comment]
> I suppose I could have been clearer -- I expected the source directory to be the current one I'm in. I don't have a hard reason why it should be that way, but that's what I naturally expect when not explicitly specifying a directory name to use.

I guess this is fine if you're doing `dagger init --sdk` or `dagger init && dagger develop --sdk` (btw, IMO the fact that both `dagger init` and `dagger develop` have the `sdk` flag feels weird to me). Now, if you're doing `dagger init` in an existing project,  I personally find it nicer about the fact that dagger creates an additional `dagger` folder to avoid any possible collisions / corruptions with files already existing in that folder. Having said that, this last inconvenience could be improved by erring and then hinting the user to add a flag for this. 

I personally don't mind adding `--source .` if I want my module'so source to be in the same folder as the dagger.json but I agree that it'd be nice if defaults could be improved. I think it boils down to what Mark is saying here: 
https://github.com/dagger/dagger/issues/6963#issuecomment-2073664141

**Comment by shykes** [comment]
FYI I made a new proposal, which would deprecate this one: https://github.com/dagger/dagger/issues/7199

**Comment by TomChv** [comment]
I removed the `sdk/typescript` label since it's not directly linked to typescript.

Also @shykes since you made a new proposal, should we close this issue?

**Comment by shykes** [comment]
This is back on my radar, since the wider #7199 was closed in favor of #7647, which is more narrowly scoped and does not address this.

FYI @sagikazarmark @jedevc @helderco @sipsma @marcosnils @nipuna-perera . Sorry that we've been slow to find the right fix. These changes to file layouts are one-way doors, really hard to correct after the fact, so better to "measure twice, cut once" as they say. But I would love to cut soon!

**Comment by shykes** [comment]
OK friends, it's finally happening! https://github.com/dagger/dagger/pull/7981
